tests/xtro-sharpie/ReleaseAttributeCheck.cs

```diff
@@ -0,0 +1,127 @@
+//
+
+using System;
+using System.Collections.Generic;
+
+using Mono.Cecil;
+
+using Clang.Ast;
+
+namespace Extrospection {
+
+	public class ReleaseAttributeCheck : BaseVisitor {
+
+		// most selectors will be found in [Export] attributes
+		public override void VisitManagedMethod (MethodDefinition method)
+		{
+			// Don't care about methods that don't have [Export] attributes
+			if (!method.HasCustomAttributes)
+				return;
+
+			// We don't care about 'void' functions
+			if (method.ReturnType.FullName == "System.Void")
+				return;
+
+			// Value types can't need '[return: Release]'
+			if (method.ReturnType.IsValueType)
+				return;
+
+			string family = null;
+			string selector = null;
+			bool hasReleaseAttribute = false;
+
+			if (method.MethodReturnType.HasCustomAttributes) {
+				foreach (var ca in method.MethodReturnType.CustomAttributes) {
+					switch (ca.Constructor.DeclaringType.Name) {
+					case "ReleaseAttribute":
+						hasReleaseAttribute = true;
+						break;
+					}
+				}
+			}
+
+			foreach (var ca in method.CustomAttributes) {
+				switch (ca.Constructor.DeclaringType.Name) {
+				case "ExportAttribute":
+					selector = (string) ca.ConstructorArguments [0].Value;
+
+					// We need to compute the selector's method family
+					// https://clang.llvm.org/docs/AutomaticReferenceCounting.html#method-families
+
+					// A selector is in a certain selector family if ignoring any leading underscore the first component of the selector either consists entirely
+					// of the name of the method family or it begins with that name followed by a character other than a lowercase letter
+					var firstLetter = 0;
+					var firstNonLowercaseLetter = selector.Length;
+					for (var i = 0; i < selector.Length; i++) {
+						var c = selector [i];
+
+						if (firstLetter == i && c == '_') {
+							// ... ignoring any leading underscores ...
+							firstLetter++;
+						}  else if (c < 'a' || c > 'z') {
+							firstNonLowercaseLetter = i;
+							break;
+						}
+					}
+					family = selector.Substring (0, firstNonLowercaseLetter - firstLetter);
+					break;
+				}
+			}
+
+			switch (family) {
+			case "init": // in many cases we have custom init/constructor code, which seems to be correct, so ignore the 'init' family for now.
+				break;
+			case "alloc":
+			case "copy":
+			case "mutableCopy":
+			case "new":
+				if (!hasReleaseAttribute) {
+					var framework = Helpers.GetFramework (method);
+					Log.On (framework).Add ($"!missing-release-attribute-on-return-value! {method.FullName}'s selector's ('{selector}') Objective-C method family ('{family}') indicates that the native method returns a retained object, and as such a '[return: Release]' attribute is required.");
+				}
+				break;
+			default:
+				break;
+			}
+
+		}
+
+		// We should also look at the native definition for family attributes: __attribute__((objc_method_family(...))
```

Comment:
filed an issue ?

Reply:
no, I'm already fixing it in Objective-Sharpie, but it takes a while to get it in and released, so I thought I should just get the rest in first.